recipe-client-addon/content/shield-content-frame.js

```diff
@@ -0,0 +1,130 @@
+ * @implements EventListener
+ */
+class ShieldFrameListener {
+  handleEvent(event) {
+    // Abort if the current page isn't about:studies.
+    if (!this.ensureTrustedOrigin()) {
+      return;
+    }
+
+    // We waited until after we received an event to register message listeners
+    // in order to save resources for tabs that don't ever load about:studies.
+    addMessageListener("Shield:ShuttingDown", this);
+    addMessageListener("Shield:ReceiveStudyList", this);
+
+    switch (event.detail.action) {
+      // Actions that require the parent process
+      case "GetRemoteValue:StudyList":
+        sendAsyncMessage("Shield:GetStudyList");
+        break;
+      case "RemoveStudy":
+        sendAsyncMessage("Shield:RemoveStudy", event.detail.data);
+        break;
+      // Actions that can be performed in the content process
+      case "GetRemoteValue:ShieldLearnMoreHref":
+        this.triggerPageCallback(
+          "ReceiveRemoteValue:ShieldLearnMoreHref",
+          frameGlobal.AboutPages.aboutStudies.getShieldLearnMoreHref()
+        );
+        break;
+      case "NavigateToDataPreferences":
+        this.navigateToDataPreferences();
+        break;
+    }
+  }
+
+  /**
+   * Check that the current webpage's origin is about:studies.
+   * @return {Boolean}
+   */
+  ensureTrustedOrigin() {
+    return content.document.documentURI.startsWith("about:studies");
+  }
+
+  /**
+   * Handle messages from the parent process.
+   * @param {Object} message
+   *   See the nsIMessageListener docs.
+   */
+  receiveMessage(message) {
+    switch (message.name) {
+      case "Shield:ReceiveStudyList":
+        this.triggerPageCallback("ReceiveRemoteValue:StudyList", message.data.studies);
+        break;
+      case "Shield:ShuttingDown":
+        this.onShutdown();
+        break;
+    }
+  }
+
+  /**
+   * Trigger an event to communicate with the unprivileged about: page.
+   * @param {String} type
+   * @param {Object} detail
+   */
+  triggerPageCallback(type, detail) {
+    // Do not communicate with untrusted pages.
+    if (!this.ensureTrustedOrigin()) {
+      return;
+    }
+
+    // Clone details and use the event class from the unprivileged context.
+    const event = new content.document.defaultView.CustomEvent(type, {
+      bubbles: true,
+      detail: Cu.cloneInto(detail, content.document.defaultView),
+    });
+    content.document.dispatchEvent(event);
+  }
+
+  onShutdown() {
+    removeMessageListener("Shield:SendStudyList", this);
+    removeMessageListener("Shield:ShuttingDown", this);
+    removeEventListener("Shield", this);
+  }
+
+  navigateToDataPreferences() {
+    if (Services.prefs.getBoolPref(USE_OLD_PREF_ORGANIZATION_PREF)) {
+      content.location = "about:preferences#advanced";
+      addEventListener("load", () => {
+        const tabbox = content.document.getElementById("advancedPrefs");
+        const tab = content.document.getElementById("dataChoicesTab");
+        tabbox.selectedTab = tab;
```

Comment:
So... the tests show this works, right, even with e10s enabled? Meanwhile, though, I am very confused, because about:preferences opens in the parent process, but this page opens in the child process. While the same script text will be loaded, because they run in different processes, I'm confused how this load listener would fire for the load that's happening in the parent process, and how the `content.document` reference would point to that thing. Can you doublecheck `about:studies` is really loading in the content process? Checking with the browser console for `gBrowser.selectedBrowser.isRemoteBrowser` with e10s turned on and about:studies loaded in the current tab should work (ie return `true`).